src/Components/Components/src/ErrorBoundaryBase.cs

```diff
@@ -0,0 +1,126 @@
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components
+{
+    /// <summary>
+    /// A base class for error boundary components.
+    /// </summary>
+    public abstract class ErrorBoundaryBase : ComponentBase, IErrorBoundary
+    {
+        private int _errorCount;
+
+        /// <summary>
+        /// The content to be displayed when there is no error.
+        /// </summary>
+        [Parameter] public RenderFragment? ChildContent { get; set; }
+
+        /// <summary>
+        /// The content to be displayed when there is an error.
+        /// </summary>
+        [Parameter] public RenderFragment<Exception>? ErrorContent { get; set; }
+
+        /// <summary>
+        /// The maximum number of errors that can be handled. If more errors are received,
+        /// they will be treated as fatal. Calling <see cref="Recover"/> resets the count.
+        /// </summary>
+        [Parameter] public int MaximumErrorCount { get; set; } = 100;
+
+        /// <summary>
+        /// Gets the current exception, or null if there is no exception.
+        /// </summary>
+        protected Exception? CurrentException { get; private set; }
+
+        /// <summary>
+        /// Resets the error boundary to a non-errored state. If the error boundary is not
+        /// already in an errored state, the call has no effect.
+        /// </summary>
+        public void Recover()
+        {
+            if (CurrentException is not null)
+            {
+                _errorCount = 0;
+                CurrentException = null;
+                StateHasChanged();
+            }
+        }
+
+        /// <summary>
+        /// Invoked by the base class when an error is being handled. Typically, derived classes
+        /// should log the exception from this method.
+        /// </summary>
+        /// <param name="exception">The <see cref="Exception"/> being handled.</param>
+        protected abstract Task OnErrorAsync(Exception exception);
+
+        void IErrorBoundary.HandleException(Exception exception)
+        {
+            if (exception is null)
+            {
+                // This would be a framework bug if it happened. It should not be possible.
+                throw new ArgumentNullException(nameof(exception));
+            }
+
+            // If rendering the error content itself causes an error, then re-rendering on error risks creating an
+            // infinite error loop. Unfortunately it's very hard to distinguish whether the error source is "child content"
+            // or "error content", since the exceptions can be received asynchronously, arbitrarily long after we switched
+            // between normal and errored states. Without creating a very intricate coupling between ErrorBoundaryBase and
+            // Renderer internals, the obvious options are either:
```

Comment:
We could "easily" determine this within the renderer and pass that info to the handler, isn't it?

The callsite for HandleException knows the handler and the source, so it can just compare them for reference equality and pass that information to handle exception (I think that's likely useful).

If we were todo that, then do we still need a limit on the number of exceptions? We could choose to say something like:
* If the exception is not within the ErrorBoundary component handle it, otherwise throw?

Reply:
It's not at all easy, as far as I know.

> If the exception is not within the ErrorBoundary component handle it, otherwise throw?

Two counterexamples:

1. If the ErrorBoundary is not yet in an errored state (or has recovered), then the exceptions will be within it, and should not be fatal
2. Even if the exception appears not to be from within the ErrorBoundary, that might be because some intermediate child has since been disposed, and in that case it might still be from the error content

Reply:
Fair enough, I don't have a deep enough intuition about this, so I'll trust your judgment (and the tests) here.